### DISCUSSION TITLE
FastAPI scoring server in MLFLOW

### DISCUSSION THREAD

**Discussion by shubh0508** [mlflow/mlflow#5601] [Ideas]
Hi Team, I have a use case where I want to deploy ML Models with low latency and high throughput. 
In MLFlow, I couldn't achieve it, so I tried FastAPI and it showed quite good results. 

Thus, I tried to replace the Flask engine with [FastAPI in MLFlow](https://github.com/mlflow/mlflow/pull/5599).  But I am getting very low throughput per process on using FastAPI in MLFlow as compared to standalone FastAPI serving for the same model.
Vegeta benchmarking results:
`
MLFlow with FastAPI Test1: 
(base) [ec2-user@ip-172-31-43-232 testing]$  cat target_all.txt | ./vegeta attack -duration=1m -rate=100  -max-workers=20 | ./vegeta report
Requests      [total, rate, throughput]         1174, 100.09, 100.07
Duration      [total, attack, wait]             11.732s, 11.73s, 2.03ms
Latencies     [min, mean, 50, 90, 95, 99, max]  1.71ms, 2.23ms, 1.87ms, 3.394ms, 3.911ms, 7.58ms, 9.813ms
Bytes In      [total, mean]                     111530, 95.00
Bytes Out     [total, mean]                     224234, 191.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:1174  

MLFlow with FastAPI Test2: 
(base) [ec2-user@ip-172-31-43-232 testing]$  cat target_all.txt | ./vegeta attack -duration=1m -rate=200  -max-workers=20 | ./vegeta report
Requests      [total, rate, throughput]         2159, 158.75, 157.30
Duration      [total, attack, wait]             13.725s, 13.6s, 125.605ms
Latencies     [min, mean, 50, 90, 95, 99, max]  10.256ms, 124.783ms, 125.951ms, 129.294ms, 133.891ms, 148.876ms, 293.218ms
Bytes In      [total, mean]                     209423, 97.00
Bytes Out     [total, mean]                     412369, 191.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:2159  

Standalone FastAPI Test3:
  Requests      [total, rate, throughput]         36000, 600, 599
  Latencies     [min, mean, 50, 90, 95, 99, max]  1.515ms, 2.608ms, 2.734ms, 3.386ms, 3.611ms, 4.204ms, 4.795ms

`

On using FastAPI, I am getting throughput to be more than 600 RPS per Gunicorn worker, while on using FastAPI with MLFlow, I am getting 100 RPS at max without affecting the response time of the model. 
I think I am missing something in this [PR](https://github.com/mlflow/mlflow/pull/5599). Can someone please help?

**Comment by kinthaiofficial** [comment]
Using MLflow with FastAPI for a scoring server is a solid production pattern. Here's the implementation that works well:

**Basic scoring server:**
```python
from fastapi import FastAPI
from pydantic import BaseModel
import mlflow
import mlflow.pyfunc
import numpy as np

app = FastAPI()

# Load model at startup
model = mlflow.pyfunc.load_model("models:/my-model/Production")

class PredictRequest(BaseModel):
    data: list[list[float]]

@app.post("/predict")
async def predict(request: PredictRequest):
    input_df = pd.DataFrame(request.data)
    predictions = model.predict(input_df)
    return {"predictions": predictions.tolist()}

@app.get("/health")
async def health():
    return {"status": "ok", "model": "my-model", "stage": "Production"}
```

**For production:**

**Async model loading:**
If the model is large, loading it in the FastAPI lifespan context (`@asynccontextmanager`) prevents the server from accepting requests before the model is ready:

```python
@asynccontextmanager
async def lifespan(app: FastAPI):
    app.state.model = mlflow.pyfunc.load_model(MODEL_URI)
    yield
    # cleanup if needed

app = FastAPI(lifespan=lifespan)
```

**Model version management:**
Don't hardcode `Production` stage — use the run ID or version for reproducibility in production:
```python
MODEL_URI = os.getenv("MODEL_URI", "models:/my-model/Production")
```

**Batch prediction vs single:**
For throughput, accept batches: `data: list[list[float]]` → predict all at once → return all predictions. Single-item prediction is expensive at scale.

**Monitoring:**
Log `mlflow.log_metric()` on each prediction for input distribution drift detection. Or use a dedicated drift detection library (Evidently, Whylogs).

What scale are you targeting — requests/second?

**Comment by SahilKumar75** [comment]
The throughput gap you're seeing isn't a FastAPI issue — it's the MLflow model wrapper overhead. When MLflow serves a model via `mlflow.pyfunc`, every prediction goes through schema enforcement, input validation, and type coercion on top of your model's actual inference. Standalone FastAPI skips all of that.

A few things that actually help:

**1. Disable input schema enforcement**

MLflow validates input against the model signature on every request. If you logged your model with a signature, that validation runs every time. You can skip it by loading the model's underlying flavor directly instead of going through `pyfunc`:

```python
import mlflow.sklearn  # or mlflow.xgboost, etc.
model = mlflow.sklearn.load_model("models:/my-model/Production")
```

Then call `model.predict()` directly. No MLflow wrapper overhead.

**2. The Gunicorn worker config matters**

MLflow's built-in server uses sync Gunicorn workers by default. In your PR, if you're still using sync workers with FastAPI (which is async), you're losing the concurrency benefit. Use `uvicorn` workers:

```
gunicorn app:app -w 4 -k uvicorn.workers.UvicornWorker
```

**3. Batch at the request level**

The benchmark shows 100 RPS cap regardless of load — that's a worker saturation issue, not a FastAPI limit. Accept batched inputs and run a single `model.predict()` per request instead of one per item.

The real ceiling for MLflow-wrapped serving is the `pyfunc` dispatch layer. If latency is critical, load the underlying model directly and use MLflow only for tracking/registry, not serving.
